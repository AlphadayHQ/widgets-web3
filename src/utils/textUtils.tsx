import { v4 as uuidv4 } from "uuid";
// import { TTag } from "../types"; TODO - define TTag 

const URL_REGEX = /(https?:\/\/[^ ]*)/gi;

export const slugify = (text: string): string =>
  text
    .toLowerCase()
    .trim()
    .replace(/ /g, "-")
    .replace(/[^\w-]+/g, "");

/**
 * Returns a shorter from of a given string
 * @param text - string to be shortened
 * @param count - no of characters to be returned
 * @param ellipsisPos - position of ellipsis, center or end
 *
 * @returns string
 */
export const truncateWithEllipsis = (
  text: string,
  count?: number,
  ellipsisPos: "end" | "center" = "center"
): string => {
  const maxLen = count !== undefined ? count : 12;

  if (text.length > maxLen) {
    if (ellipsisPos === "end")
      return `${text.substr(0, Math.floor(maxLen))}...`;
    return `${text.substr(0, Math.floor(maxLen / 2))}...${text.substr(
      text.length - Math.floor(maxLen / 2)
    )}`;
  }
  return text;
};

export const parseNewLine = (s: string, tabCount = 2): JSX.Element => {
  return (
    <span>
      {s.split("\n").map((e, i, self) => (
        <span key={uuidv4()}>
          {e}
          {i < self.length - 1 && (
            <>
              {[...Array(tabCount)].map(() => (
                <br key={uuidv4()} />
              ))}
            </>
          )}
        </span>
      ))}
    </span>
  );
};

export const isURL = (text: string): boolean => URL_REGEX.test(text);

/**
 * Parses urls and new line elements in text of type string
 * to html format
 *
 * @returns a JSX.Element
 */
export const parseText = (text: string, tabCount = 2): JSX.Element => {
  // eslint-disable-next-line @typescript-eslint/prefer-regexp-exec
  const urls = text.match(URL_REGEX);

  if (urls) {
    return (
      <>
        {text.split(" ").map((str) =>
          urls.includes(str) ? (
            <a
              href={str}
              className="link" // className link is used to style this item from a parent container
              target="_blank"
              rel="noreferrer"
              key={uuidv4()}
            >
              {truncateWithEllipsis(str, 18, "end")}
            </a>
          ) : (
            <span key={uuidv4()}>{parseNewLine(str)} </span>
          )
        )}
      </>
    );
  }

  return parseNewLine(text, tabCount);
};

// TODO TTag definition missing
// export const textHasTags = (text: string, tags: TTag[]): boolean =>
//   text.split(" ").some((curr) =>
//     tags.some((tag) => {
//       const filterTagName = new RegExp(tag.name, "i");
//       const filterTagSlug = new RegExp(tag.slug, "i");
//       return filterTagName.test(curr) || filterTagSlug.test(curr);
//     })
//   );
