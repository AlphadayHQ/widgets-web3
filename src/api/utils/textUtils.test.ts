import renderer from "react-test-renderer";
import { parseText, slugify } from "./textUtils";

describe("Test for text Utilities", () => {
    it("check the slug is correctly formatted", () => {
        const slugTests = [
            { test: " bitcoin ", result: "bitcoin" },
            { test: "terra luna", result: "terra-luna" },
            { test: "Bitcoin Cash", result: "bitcoin-cash" },
        ];
        slugTests.forEach(({ test, result }) => {
            expect(slugify(test)).toBe(result);
        });
    });

    it("checks jsx.Element is created correctly from a string by parseText", () => {
        const rawText =
            "Yellow Fever:\n\nRequired if arriving from \n- Angola\n- Peru\n- Dominican Republic. \n\n Visit https://www.who.int/publications-detail-redirect/9789241563871 or https://www.sciencedirect.com/topics/immunology-and-microbiology/infectious-diseases to learn more.";
        const tree = renderer.create(parseText(rawText)).toJSON();
        expect(tree).toMatchSnapshot();
    });
});
