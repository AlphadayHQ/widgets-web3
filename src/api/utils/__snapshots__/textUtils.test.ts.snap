// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test for text Utilities checks jsx.Element is created correctly from a string by parseText 1`] = `
Array [
  <span>
    <span>
      <span>
        Yellow
      </span>
    </span>
     
  </span>,
  <span>
    <span>
      <span>
        Fever:
        <br />
        <br />
      </span>
      <span>
        <br />
        <br />
      </span>
      <span>
        Required
      </span>
    </span>
     
  </span>,
  <span>
    <span>
      <span>
        if
      </span>
    </span>
     
  </span>,
  <span>
    <span>
      <span>
        arriving
      </span>
    </span>
     
  </span>,
  <span>
    <span>
      <span>
        from
      </span>
    </span>
     
  </span>,
  <span>
    <span>
      <span>
        <br />
        <br />
      </span>
      <span>
        -
      </span>
    </span>
     
  </span>,
  <span>
    <span>
      <span>
        Angola
        <br />
        <br />
      </span>
      <span>
        -
      </span>
    </span>
     
  </span>,
  <span>
    <span>
      <span>
        Peru
        <br />
        <br />
      </span>
      <span>
        -
      </span>
    </span>
     
  </span>,
  <span>
    <span>
      <span>
        Dominican
      </span>
    </span>
     
  </span>,
  <span>
    <span>
      <span>
        Republic.
      </span>
    </span>
     
  </span>,
  <span>
    <span>
      <span>
        <br />
        <br />
      </span>
      <span>
        <br />
        <br />
      </span>
      <span />
    </span>
     
  </span>,
  <span>
    <span>
      <span>
        Visit
      </span>
    </span>
     
  </span>,
  <a
    className="link"
    href="https://www.who.int/publications-detail-redirect/9789241563871"
    rel="noreferrer"
    target="_blank"
  >
    https://www.who.in...
  </a>,
  <span>
    <span>
      <span>
        or
      </span>
    </span>
     
  </span>,
  <a
    className="link"
    href="https://www.sciencedirect.com/topics/immunology-and-microbiology/infectious-diseases"
    rel="noreferrer"
    target="_blank"
  >
    https://www.scienc...
  </a>,
  <span>
    <span>
      <span>
        to
      </span>
    </span>
     
  </span>,
  <span>
    <span>
      <span>
        learn
      </span>
    </span>
     
  </span>,
  <span>
    <span>
      <span>
        more.
      </span>
    </span>
     
  </span>,
]
`;
